### PR TITLE
✨ amp-subscriptions: Allow score factors to control rendering

### LINKED ISSUE #23621: amp-subscriptions: support conditional rendering based on scoring factors
## Describe the new feature or change to an existing feature you'd like to see

Add the ability to conditionally display dialogs and buttons based on items currently used in scoring factors, such as with amp-subscriptions-google "ready to pay"; example, if user is ready to pay, show button, if not, don't show button.

## Describe alternatives you've considered

N/A
## Additional context

N/A

### PR DESCRIPTION
**Pull request by jpettitt**
# Instructions:

Fixes #23621

### CODE REVIEW

File: extensions/amp-subscriptions/0.1/platform-store.js [hunk 1]
```diff
@@ -242,6 +242,58 @@ export class PlatformStore {
...
   }
 
+  /**
+   * Get scoreFactor states for each platform
+   * @return {Promise<!JsonObject>}
+   *
+   * return value looks somethinglike this
+   * {
+   *    isReadyToPay: {
+   *     'subscription-google-com': 1,
```
Comment (dvoytenko) [inline extensions/amp-subscriptions/0.1/platform-store.js]: This should be "subscribe-google-com". But do you think maybe the `platform -> score` will be more expected? E.g.

```
{
  'subscribe-google-com':  {
    isReadyToPay: 1,
    supportsViewer: 1,
    ...
  }
}
```

File: extensions/amp-subscriptions/0.1/platform-store.js [hunk 2]
```diff
@@ -242,6 +242,58 @@ export class PlatformStore {
...
+   *     local: 0,
+   *    },
+   *    supportdViewer: {
```
Comment (dvoytenko) [inline extensions/amp-subscriptions/0.1/platform-store.js]: "supportsViewer"

File: extensions/amp-subscriptions/0.1/platform-store.js [hunk 3]
```diff
@@ -242,6 +242,58 @@ export class PlatformStore {
...
+   *     'subscription-google-com': 1,
```
Comment (dvoytenko) [inline extensions/amp-subscriptions/0.1/platform-store.js]: "subscribe-"

File: extensions/amp-subscriptions/0.1/platform-store.js [hunk 4]
```diff
@@ -242,6 +242,58 @@ export class PlatformStore {
...
+    const states = dict({});
+    return Promise.all(
+      Object.values(SubscriptionsScoreFactor).map(scoreFactor => {
+        states[scoreFactor] = dict();
+        return Promise.all(
+          this.serviceIds_.map(platformId =>
+            this.getScoreFactorPromiseFor_(platformId, scoreFactor).then(
+              factorValue => {
+                states[scoreFactor][
+                  platformId.replace(/\./g, '_')
```
Comment (dvoytenko) [inline extensions/amp-subscriptions/0.1/platform-store.js]: Dash or underscore? Above it documents "dash".

File: extensions/amp-subscriptions/0.1/service-adapter.js [hunk 5]
```diff
@@ -58,6 +58,14 @@ export class ServiceAdapter {
     return this.subscriptionService_.getAnalytics();
   }
 
+  /**
+   * gets Score factors for all platforms
+   * @return {Promise<!JsonObject>}
```
Comment (dvoytenko) [inline extensions/amp-subscriptions/0.1/service-adapter.js]: `!` or `?`

File: extensions/amp-subscriptions/0.1/test/test-local-subscriptions.js [hunk 6]
```diff
@@ -56,12 +56,25 @@ describes.fakeWin('LocalSubscriptionsPlatform', {amp: true}, env => {
   };
   const authUrl = configAuthUrl.replace('READER_ID', readerId);
   const pingbackUrl = configPingbackUrl.replace('READER_ID', readerId);
+  const fakeScoreStates = {
+    isReadyToPay: {
+      'subscription-google-com': 1,
```
Comment (dvoytenko) [inline extensions/amp-subscriptions/0.1/test/test-local-subscriptions.js]: "subscribe-"

File: extensions/amp-subscriptions/0.1/test/test-local-subscriptions.js [hunk 7]
```diff
@@ -56,12 +56,25 @@ describes.fakeWin('LocalSubscriptionsPlatform', {amp: true}, env => {
...
+      local: 0,
+    },
+    supportdViewer: {
```
Comment (dvoytenko) [inline extensions/amp-subscriptions/0.1/test/test-local-subscriptions.js]: "supports"

File: examples/amp-subscriptions-rtp.amp.html [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+      </div>
+    </article>
+  </main>
+
+  <footer>
+    <div class="brand-logo">PublisherLogo</div>
+  </footer>
+
+  <template type="amp-mustache" subscriptions-dialog 
+  subscriptions-display="data.foo" "NOT platforms.subscriptions.google.com.isReadyToPay">
```
[the remark this answers is not in the document]
Comment (jpettitt) [reply examples/amp-subscriptions-rtp.amp.html]: `access-expr-impl` thros if I use the `['local']` syntax.
image

File: examples/amp-subscriptions-rtp.amp.html [hunk 9]
```diff
@@ -0,0 +1,399 @@
...
+  </template>
+  <template type="amp-mustache" subscriptions-dialog subscriptions-display="NOT loggedIn">
+    <div>Dialog for non logged in users<div>
+    <!-- Shows a Subscribe with Google button if the user is ready to pay -->
+    <button subscriptions-display="scores.isReadyToPay.subscribe_google_com"
+      subscriptions-action="subscribe"
+      subscriptions-service="subscribe.google.com"
+      subscriptions-decorate>Subscribe with Google</button>
+    <!-- Shows a Publisher button if the user is not ready to pay -->
+    <button subscriptions-display="NOT scores.isReadyToPay.subscribe_google_com"
```
[the remark this answers is not in the document]
Comment (jpettitt) [reply examples/amp-subscriptions-rtp.amp.html]: ?

File: extensions/amp-subscriptions/0.1/platform-store.js (hunk 1 above)
Comment (jpettitt) [reply extensions/amp-subscriptions/0.1/platform-store.js]: Don't feel strongly either way. Easy enough to change. WDYT?

File: extensions/amp-subscriptions/0.1/amp-subscriptions.js [hunk 10]
```diff
@@ -662,6 +662,14 @@ export class SubscriptionService {
         return getValueForExpr(entitlement.json(), field);
       });
   }
+
+  /**
+   * gets Score factors for all platforms
```
[the remark this answers is not in the document]
Comment (jpettitt) [reply extensions/amp-subscriptions/0.1/amp-subscriptions.js]: fixed

File: extensions/amp-subscriptions/0.1/amp-subscriptions.js [hunk 11]
```diff
@@ -662,6 +662,14 @@ export class SubscriptionService {
...
+   * @return {Promise<!JsonObject>}
```
[the remark this answers is not in the document]
Comment (jpettitt) [reply extensions/amp-subscriptions/0.1/amp-subscriptions.js]: fixed

File: extensions/amp-subscriptions/0.1/platform-store.js [hunk 12]
```diff
@@ -242,6 +242,58 @@ export class PlatformStore {
     return this.entitlementDeferredMap_[serviceId].promise;
   }
 
+  /**
+   * Get scoreFactor states for each platform
+   * @return {Promise<!JsonObject>}
```
[the remark this answers is not in the document]
Comment (jpettitt) [reply extensions/amp-subscriptions/0.1/platform-store.js]: fixed

File: extensions/amp-subscriptions/0.1/platform-store.js (hunk 2 above)
Comment (jpettitt) [reply extensions/amp-subscriptions/0.1/platform-store.js]: fixed

File: extensions/amp-subscriptions/0.1/platform-store.js (hunk 3 above)
Comment (jpettitt) [reply extensions/amp-subscriptions/0.1/platform-store.js]: fixed

File: extensions/amp-subscriptions/0.1/platform-store.js (hunk 4 above)
Comment (jpettitt) [reply extensions/amp-subscriptions/0.1/platform-store.js]: underscore, the expression parser throws on dash

File: extensions/amp-subscriptions/0.1/service-adapter.js (hunk 5 above)
Comment (jpettitt) [reply extensions/amp-subscriptions/0.1/service-adapter.js]: fixed

File: extensions/amp-subscriptions/0.1/test/test-local-subscriptions.js (hunk 6 above)
Comment (jpettitt) [reply extensions/amp-subscriptions/0.1/test/test-local-subscriptions.js]: fixed

File: extensions/amp-subscriptions/0.1/test/test-local-subscriptions.js (hunk 7 above)
Comment (jpettitt) [reply extensions/amp-subscriptions/0.1/test/test-local-subscriptions.js]: fixed

File: examples/amp-subscriptions-rtp.amp.html (hunk 8 above)
[the remark this answers is not in the document]
Comment (dvoytenko) [reply examples/amp-subscriptions-rtp.amp.html]: We'd certainly need to extend expression grammar a bit. I was asking if we'd benefit for that kind of extension. I'll be happy to make changes if that helps you.

File: examples/amp-subscriptions-rtp.amp.html (hunk 9 above)
[the remark this answers is not in the document]
Comment (dvoytenko) [reply examples/amp-subscriptions-rtp.amp.html]: I meant that we have both `scores.subscribe_google_com.isReadyToPay` and `scores.isReadyToPay.subscribe_google_com`?

File: extensions/amp-subscriptions/0.1/platform-store.js (hunk 1 above)
Comment (dvoytenko) [reply extensions/amp-subscriptions/0.1/platform-store.js]: I think it depends on whether we'd also do something like `scores.isReadyToPay` resolving to something for "all" platforms. If not - I think `scores.platform.isReadyToPay` seems better to me.

File: extensions/amp-subscriptions/0.1/platform-store.js (hunk 4 above)
Comment (dvoytenko) [reply extensions/amp-subscriptions/0.1/platform-store.js]: I could also support "dash" in names if that helps. There's a lot of precedent for dashed names in HTML/XML and family.

Comment (jpettitt) [reply extensions/amp-subscriptions/0.1/platform-store.js]: Dash would be good too, although supporting the `['name.with.dots']` format will remove that need

**Pull request merged**